crates/ra_syntax/src/grammar/expressions.rs

```diff
@@ -54,8 +57,9 @@ pub(crate) fn block(p: &mut Parser) {
             _ => {
                 // test block_items
                 // fn a() { fn b() {} }
-                let has_attrs = p.at(POUND);
                 let m = p.start();
+                let has_attrs = p.at(POUND);
+                attributes::outer_attributes(p);
                 match items::maybe_item(p, items::ItemFlavor::Mod) {
```

Comment:
... and move it over here as `if p.at(LET_KW) => let_stmt, else maybe_item`. 

In general, it's better to try to arrange control flow in such way thatn every thing is parsed in exactly one place.